File: src/leetcodejava/array/Offer11.java
package leetcodejava.array;


import org.junit.Test;

/**
 * This is the solution of No.offer 11 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 把一个数组最开始的若干个元素搬到数组的末尾，我们称之为数组的旋转。输入一个递增排序的数组的一个旋转，输出旋转数组的最小元素。例如，数组 [3,4,5,1,2] 为 [1,2,3,4,5] 的一个旋转，该数组的最小值为1。  
 * <p>
 * 示例 1：
 * <p>
 * 输入：[3,4,5,1,2]
 * 输出：1
 * 示例 2：
 * <p>
 * 输入：[2,2,2,0,1]
 * 输出：0
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
public class Offer11 {

    @Test
    public void offerTest() {
        int numbers[] = {3, 4, 5, 1, 2};
        int i = minArray(numbers);
        System.out.println(i);
    }

    /**
     * 旋转数组的最小数字
     *
     * @param numbers 数组
     * @return 最小数
     */
    public int minArray(int[] numbers) {
        int l = 0, r = numbers.length - 1;
        while (l < r) {
            int mid = ((r - l) >> 1) + l;
            //只要右边比中间大，那右边一定是有序数组
            if (numbers[r] > numbers[mid]) {
                r = mid;
            } else if (numbers[r] < numbers[mid]) {
                l = mid + 1;
                //去重
            } else {
                r--;
            }
        }
        return numbers[l];
    }
}

File: src/leetcodepython/array/offer_11.py
# -*- coding:utf-8 -*-
'''
/**
 * This is the solution of No.offer 11 problem in the LeetCode,
 * the website of the problem is as follow:
 * https://leetcode-cn.com/problems/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof
 * <p>
 * The description of problem is as follow:
 * ==========================================================================================================
 * 把一个数组最开始的若干个元素搬到数组的末尾，我们称之为数组的旋转。输入一个递增排序的数组的一个旋转，输出旋转数组的最小元素。
 * 例如，数组 [3,4,5,1,2] 为 [1,2,3,4,5] 的一个旋转，该数组的最小值为1。  
 * <p>
 * 示例 1：
 * <p>
 * 输入：[3,4,5,1,2]
 * 输出：1
 * 示例 2：
 * <p>
 * 输入：[2,2,2,0,1]
 * 输出：0
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/xuan-zhuan-shu-zu-de-zui-xiao-shu-zi-lcof
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 * ==========================================================================================================
 *
 * @author zhangyu (zhangyuyu417@gmail.com)
 */
'''
from typing import List


class Solution:
    def min_array(self, nums: List[int]) -> int:
        '''
            找到最小数字
        Args:
            nums: 数组
            target: 目标值
        Returns:
            数组下标
        '''
        l = 0
        r = len(nums) - 1
        while l < r:
            mid = (l + r) // 2
            if nums[mid] < nums[r]:
                r = mid
            elif nums[mid] > nums[r]:
                l = mid + 1
            else:
                r -= 1
        return nums[l]


if __name__ == '__main__':
    nums = [3, 4, 5, 1, 2]
    solution = Solution()
    result = solution.min_array(nums)
    print(result)
    assert result == 1
